src/rt-app.c

```diff
@@ -415,6 +434,72 @@ static int run_event(event_data_t *event, int dry_run,
 			pthread_yield();
 		}
 		break;
+	case rtapp_fork:
+		{
+			log_debug("fork");
+
+			/*
+			 * If multiple threads race to fork, we must ensure
+			 * each one sees a unique index. Hence the lock.
+			 */
+			pthread_mutex_lock(&fork_mutex);
+			int new_thread_index = nthreads++;
+			threads = realloc(threads, nthreads * sizeof(*threads));
+			thread_data_t *tdata = malloc(sizeof(thread_data_t));
+
+			if (!threads || !tdata) {
+				log_error("Failed to allocate memory for a new fork: %s", rdata->res.fork.tdata->name);
+				pthread_mutex_unlock(&fork_mutex);
+				exit(EXIT_FAILURE);
+			}
+
+			/* Find the thread data associated with this fork */
+			if (!rdata->res.fork.tdata) {
+				rdata->res.fork.tdata = find_thread_data(rdata->res.fork.ref, &opts);
```

Comment:
worth to add a comment to mention that you keep a pointer to the tdata to prevent re doing it for next loop.
Because apart for this optim, the pointer is not needed.
This raise a question ? why not setting this pointer when parsing the json file instead of waiting for the 1st event ?

Reply:
Because we have to wait for all threads to be parsed before we can search for the reference. We could do something at parse time probably but this seems a lot more efficient and simpler to me.

I'll add comments.

Reply:
yes but we can handle that. And I think that we have to.
With the new mechanism of duplicating thread_data, we have a waste a memory and structure. 
In current version, we create a thread_data per task and use this structure directly in the thread.
With fork feature, we are duplicating the thread_structure and use the copy in the thread but we are still creating one thread_structure per instance. This means with "instance" : 1000 as an example, we will allocate a thread_data array of 1000 cells, parse 1000 times the json file to fill the structures but then duplicate them so we have 2000 structures and half of them are unused.
The right way to do would be to create only 1 thread_data structure per task node and then duplicate it when we create thread.
In addition, we can also split the parsing in 2 steps:
1st step parse all task nodes and create this main structures
2nd step, parse the phases and fill the array of phases and events. At that time, you can directly set the pointer of fork event instead of waiting for the 1st event to run

Reply:
>  parse 1000 times the json file to

This will never happen, we parse the json file once, no? I just extended parse_tasks() but didn't change the json file is parsed.

We can optimize the parsing loop to ignore instances; I'll try to spend a little time on that. But FWIW the sizeof(thread_data_t) is 160 bytes. If we spawn 4000 (which is very unlikely for the purpose of rt-app other than to create a forkbomb maybe) we will need ~640KiB. If this comment appeared earlier it would have been easier to address - but given the long time since I last touched it I agree this is suboptimal but for all purposes of rt-app this is just fine and will hardly be a limitation except for the tiniest of systems where Linux probably can't run on them anyway.

Reply:
no look at parse_tasks, we parse the task node for each instance to create the thread_data of each thread
For fork, you change the behavior and you just need to create one instance because you duplicate it for each thread. Do it completely and cleanup unused data.
I'm sorry about the delay in review but can't accept sub optimal behavior like that even if it would have been only 1 Byte per instance. 
You change the init sequence for the fork event but you have to do that completely and cleanly and not changing part of it and let dead code or data which means:
- remove useless creation of structure
- don't resolve at runtime something that could have been done during the parse of the json file. all other events setup everything during the parse of the json file so we need to be consistent. Furthermore, you create a difference between the 1st and the next events without any good reason apart from reducing the code refactoring

Reply:
No one wants to waste anything. But when the review process is very slow and disjointed - it seems more fruitful for everyone's time to accept what we have and push improvements in top so that progress is made. Especially the issue highlighted here is a no-issue from practical point of view and meets the requirement of rt-app adequately.

So once I fix this will this be merged? A recurring theme in the review process is to pop up to request some changes then disappear - after that appear again to request changes in different area and this is wasting my time as I want to address all comments in one go. If you want to look at the code again please do so and let me know so that I can address everything in one go. 

I appreciate all comments given and time taken to look at this request by the way - I don't want the above to sound ungrateful in anyway :)

Reply:
This is addressed here now:

https://github.com/scheduler-tools/rt-app/pull/53/commits/07a1138edf79c7cfe17ed347b0a198ecf7be194e